theano/tensor/tests/test_basic.py

```diff
@@ -7271,7 +7271,284 @@ def test_bad_number_of_shape(self):
             self.assertRaises(AssertionError, f, xval, shape)
 
 
+class test_diagonal(unittest.TestCase):
+    """
+    Test that tensor.diagonal has the same behavior as and numpy.diagonal.
+
+    (1) When given a matrix, numpy.diagonal returns a vector which is the
+    diagonal of the matrix.
+    (2) When given a tensor with `n>=3` dimensions, it returns an `n-1`
+    dimensional tensor extracting the desired diagonal on dimensions specified
+    by `axis1` and `axis2`. It's implemented on CPU, but on GPU we only allow
+    default settings currently.
+
+    We test the input type check, the forward computation, infer_shape, and
+    the gradient respectively. 
+    """
+    def __init__(self, name, mode=None, shared=tensor._shared,
+                 floatX=None, type=tensor.TensorType):
+        self.mode = mode
+        self.shared = shared
+        if floatX is None:
+            floatX = config.floatX
+        self.floatX = floatX
+        self.type = type
+        super(test_diagonal, self).__init__(name)
+
+    def test_diagonal_input(self):
+        # test that it accepts the right form of input
+        x = theano.tensor.matrix()
+        y = diagonal(x)
+        assert y.owner.op.__class__ == Diagonal
+
+        x = theano.tensor.tensor3()
+        y = diagonal(x)
+        assert y.owner.op.__class__ == Diagonal
+
+        # other types should raise error
+        x = theano.tensor.vector()
+        ok = False
+        try:
+            y = diagonal(x)
```

Comment:
It would be better with ```assert_raises``` like [here](https://github.com/hantek/Theano/blob/diag/theano/tensor/tests/test_basic.py#L5509)